llvm/include/llvm-c/CAS/PluginAPI_functions.h

```diff
@@ -0,0 +1,244 @@
+ * that the plugin was compiled with.
+ * Intended for assisting compatibility with different versions.
+ */
+LLCASAPI_PUBLIC void llcasapi_get_plugin_version(unsigned *major,
+                                                 unsigned *minor);
+
+/**
+ * Releases memory of C string pointers provided by other functions.
+ */
+LLCASAPI_PUBLIC void llcasapi_string_dispose(char *);
+
+LLCASAPI_PUBLIC llcasapi_cas_options_t llcasapi_cas_options_create(void);
+
+LLCASAPI_PUBLIC void llcasapi_cas_options_dispose(llcasapi_cas_options_t);
+
+/**
+ * Receives the \c LLCASAPI_VERSION_MAJOR and \c LLCASAPI_VERSION_MINOR values
+ * that the client was compiled with.
+ * Intended for assisting compatibility with different versions.
+ */
+LLCASAPI_PUBLIC void
+llcasapi_cas_options_set_client_version(llcasapi_cas_options_t, unsigned major,
+                                        unsigned minor);
+
+/**
+ * Receives a local file-system path that the plugin should use for any on-disk
+ * resources/caches.
+ */
+LLCASAPI_PUBLIC void
+llcasapi_cas_options_set_ondisk_path(llcasapi_cas_options_t, const char *path);
+
+/**
+ * Receives a name/value strings pair, for the plugin to set as a custom option
+ * it supports. These are usually passed through as invocation options and are
+ * opaque to the client.
+ *
+ * \param error optional pointer to receive an error message if an error
+ * occurred. If set, the memory it points to needs to be released via
+ * \c llcasapi_string_dispose.
+ * \returns true if there was an error, false otherwise.
+ */
+LLCASAPI_PUBLIC bool llcasapi_cas_options_set_option(llcasapi_cas_options_t,
+                                                     const char *name,
+                                                     const char *value,
+                                                     char **error);
+
+/**
+ * Creates a new \c llcasapi_cas_t object. The objects returned from the other
+ * functions are only valid to use while the \c llcasapi_cas_t object that they
+ * came from is still valid.
+ *
+ * \param error optional pointer to receive an error message if an error
+ * occurred. If set, the memory it points to needs to be released via
+ * \c llcasapi_string_dispose.
+ * \returns \c NULL if there was an error.
+ */
+LLCASAPI_PUBLIC llcasapi_cas_t llcasapi_cas_create(llcasapi_cas_options_t,
+                                                   char **error);
+
+LLCASAPI_PUBLIC void llcasapi_cas_dispose(llcasapi_cas_t);
+
+/**
+ * \returns the hash schema name that the plugin is using. The string memory it
+ * points to needs to be released via \c llcasapi_string_dispose.
+ */
+LLCASAPI_PUBLIC char *llcasapi_cas_get_hash_schema_name(llcasapi_cas_t);
+
+/**
+ * Parses the printed digest and returns the digest hash bytes.
+ *
+ * \param printed_digest a C string that was previously provided by
+ * \c llcasapi_digest_print.
+ * \param bytes pointer to a buffer for writing the digest bytes.
```

Comment:
Can this be nullptr if bytes_size is 0?  Would be good to clarify since this is an easy way to determine the size:

```
 llcasapi_digest_parse(cas, digest_string, nullptr, 0, nullptr);
```

Reply:
Added such documentation comment.